File: R/train_model.R
# =============================================================================
# SCRIPT   : train_model.R
# STUDY    : NSVT-001 | NovaStat Therapeutics (FICTIONAL - AI-generated demo)
# PURPOSE  : Train patient dropout prediction model using tidymodels
#            Random Forest + XGBoost; saves model for Shiny app and vetiver
# NOTE     : ALL DATA IS SYNTHETIC. For demonstration purposes only.
# =============================================================================

library(tidymodels)
library(tidyverse)
library(vip)
library(vetiver)

set.seed(42)

# --- Load data ---
adsl_path <- "data/adam/adsl.csv"
if (!file.exists(adsl_path)) {
  stop("ADSL not found. Run sas/generate_sdtm.sas then sas/create_adam.sas first.")
}
adsl <- read_csv(adsl_path, show_col_types = FALSE)

message(sprintf("Loaded ADSL: %d subjects | Dropout rate: %.1f%%",
                nrow(adsl), mean(adsl$DPTFL == "Y") * 100))

# =============================================================================
# FEATURE SELECTION
# ML features matching clinical predictors used in pharma dropout studies
# =============================================================================

model_data <- adsl |>
  filter(ITTFL == "Y") |>
  select(
    # Target
    DPTFL,
    # Demographics
    AGE, SEXN, RACEN, REGIONN,
    # Disease characteristics
    DAS28BL, BMIBL, TRT01PN,
    # AE burden (primary drivers of dropout)
    AENUM, AESNUM, AEMAXSEVN, N_ALT_AE, N_NEURO_GI_AE,
    # Compliance
    COMPLPCT, N_MISSED_DOSES
  ) |>
  mutate(
    DPTFL   = factor(DPTFL, levels = c("N", "Y")),
    SEXN    = factor(SEXN),
    RACEN   = factor(RACEN),
    REGIONN = factor(REGIONN),
    TRT01PN = factor(TRT01PN)
  ) |>
  drop_na()

message(sprintf("Model dataset: %d subjects | %d features", nrow(model_data), ncol(model_data) - 1))

# =============================================================================
# TRAIN / TEST SPLIT (70/30, stratified by outcome)
# =============================================================================

split  <- initial_split(model_data, prop = 0.70, strata = DPTFL)
train  <- training(split)
test   <- testing(split)

message(sprintf("Train: %d | Test: %d | Train dropout rate: %.1f%%",
                nrow(train), nrow(test), mean(train$DPTFL == "Y") * 100))

# 5-fold CV for model selection
folds <- vfold_cv(train, v = 5, strata = DPTFL)

# =============================================================================
# PREPROCESSING RECIPE
# =============================================================================

rec <- recipe(DPTFL ~ ., data = train) |>
  step_impute_median(all_numeric_predictors()) |>
  step_normalize(AGE, DAS28BL, BMIBL, COMPLPCT, AENUM, AESNUM, AEMAXSEVN,
                 N_ALT_AE, N_NEURO_GI_AE, N_MISSED_DOSES) |>
  step_dummy(SEXN, RACEN, REGIONN, TRT01PN, one_hot = FALSE)

# =============================================================================
# MODEL 1: RANDOM FOREST
# =============================================================================

rf_spec <- rand_forest(
  mtry  = tune(),
  trees = 500,
  min_n = tune()
) |>
  set_engine("ranger", importance = "impurity", seed = 42) |>
  set_mode("classification")

rf_wf <- workflow() |>
  add_recipe(rec) |>
  add_model(rf_spec)

rf_grid <- grid_regular(
  mtry(range = c(2, 8)),
  min_n(range = c(5, 25)),
  levels = 3
)

message("Tuning Random Forest (5-fold CV)...")
rf_res <- tune_grid(
  rf_wf,
  resamples = folds,
  grid      = rf_grid,
  metrics   = metric_set(roc_auc, accuracy, f_meas),
  control   = control_grid(save_pred = TRUE, verbose = FALSE)
)

rf_best <- select_best(rf_res, metric = "roc_auc")
message(sprintf("Best RF: mtry=%d, min_n=%d", rf_best$mtry, rf_best$min_n))

rf_final_wf <- finalize_workflow(rf_wf, rf_best)
rf_fit      <- last_fit(rf_final_wf, split)

# =============================================================================
# MODEL 2: XGBOOST
# =============================================================================

xgb_spec <- boost_tree(
  trees          = tune(),
  tree_depth     = tune(),
  learn_rate     = tune(),
  loss_reduction = tune()
) |>
  set_engine("xgboost", seed = 42) |>
  set_mode("classification")

xgb_wf <- workflow() |>
  add_recipe(rec) |>
  add_model(xgb_spec)

xgb_grid <- grid_latin_hypercube(
  trees(range = c(100, 500)),
  tree_depth(range = c(3, 7)),
  learn_rate(range = c(-3, -1)),
  loss_reduction(range = c(-5, 0)),
  size = 15
)

message("Tuning XGBoost (5-fold CV)...")
xgb_res <- tune_grid(
  xgb_wf,
  resamples = folds,
  grid      = xgb_grid,
  metrics   = metric_set(roc_auc, accuracy, f_meas),
  control   = control_grid(save_pred = TRUE, verbose = FALSE)
)

xgb_best    <- select_best(xgb_res, metric = "roc_auc")
xgb_final_wf <- finalize_workflow(xgb_wf, xgb_best)
xgb_fit     <- last_fit(xgb_final_wf, split)

# =============================================================================
# COMPARE MODELS ON TEST SET
# =============================================================================

rf_metrics  <- collect_metrics(rf_fit)
xgb_metrics <- collect_metrics(xgb_fit)

comparison <- bind_rows(
  rf_metrics  |> mutate(model = "Random Forest"),
  xgb_metrics |> mutate(model = "XGBoost")
) |>
  select(model, .metric, .estimate) |>
  pivot_wider(names_from = .metric, values_from = .estimate)

message("\n--- Model Comparison (Test Set) ---")
print(comparison)

# Select best model by ROC AUC
best_model_name <- comparison |>
  slice_max(roc_auc) |>
  pull(model)

best_fit <- if (best_model_name == "XGBoost") xgb_fit else rf_fit
best_wf  <- if (best_model_name == "XGBoost") xgb_final_wf else rf_final_wf

message(sprintf("\nBest model: %s (AUC = %.3f)",
                best_model_name,
                comparison |> slice_max(roc_auc) |> pull(roc_auc)))

# =============================================================================
# FIT FINAL MODEL ON FULL TRAINING DATA
# =============================================================================

final_model <- fit(best_wf, data = train)

# =============================================================================
# VARIABLE IMPORTANCE
# =============================================================================

if (best_model_name == "Random Forest") {
  vip_plot <- final_model |>
    extract_fit_parsnip() |>
    vip(num_features = 12, geom = "col", aesthetics = list(fill = "#1B4F8A")) +
    labs(title = "Random Forest - Variable Importance",
         subtitle = "NSVT-001 Patient Dropout Prediction") +
    theme_minimal()
} else {
  vip_plot <- final_model |>
    extract_fit_parsnip() |>
    vip(num_features = 12, geom = "col", aesthetics = list(fill = "#1B4F8A")) +
    labs(title = "XGBoost - Variable Importance",
         subtitle = "NSVT-001 Patient Dropout Prediction") +
    theme_minimal()
}
print(vip_plot)

# =============================================================================
# SAVE MODEL (vetiver for Posit Connect deployment)
# =============================================================================

dir.create("ml", showWarnings = FALSE)

# Save as vetiver model
v_model <- vetiver_model(
  model   = final_model,
  model_name = "nsvt001-dropout-model",
  description = paste(best_model_name, "model for NSVT-001 patient dropout prediction"),
  metadata = list(
    study    = "NSVT-001",
    sponsor  = "NovaStat Therapeutics",
    drug     = "NST-4892",
    target   = "Patient dropout (DPTFL=Y)",
    auc      = comparison |> slice_max(roc_auc) |> pull(roc_auc),
    features = names(select(model_data, -DPTFL)),
    disclaimer = "SYNTHETIC DATA - AI-generated for demonstration only"
  )
)

saveRDS(final_model, "ml/dropout_model.rds")
saveRDS(v_model,     "ml/vetiver_model.rds")
saveRDS(comparison,  "ml/model_comparison.rds")
saveRDS(split,       "ml/train_test_split.rds")
saveRDS(rec |> prep(), "ml/recipe.rds")

message(sprintf("\nModel saved to ml/dropout_model.rds"))
message(sprintf("Vetiver model saved to ml/vetiver_model.rds"))

# =============================================================================
# ROC CURVE
# =============================================================================

roc_data <- collect_predictions(best_fit)
roc_auc_val <- roc_auc(roc_data, truth = DPTFL, .pred_Y, event_level = "second")$.estimate

roc_curve(roc_data, truth = DPTFL, .pred_Y) |>
  autoplot() +
  labs(
    title    = paste(best_model_name, "- ROC Curve (Test Set)"),
    subtitle = sprintf("AUC = %.3f | Study NSVT-001 - Patient Dropout Prediction", roc_auc_val)
  ) +
  theme_minimal()

message(sprintf("\nTraining complete. Final model AUC: %.3f", roc_auc_val))
message("Run R/app.R to launch the interactive dashboard.")

File: python/train_model.py
"""
Script   : train_model.py
Study    : NSVT-001 | NovaStat Therapeutics (FICTIONAL - AI-generated demo)
Purpose  : Train patient dropout prediction model using scikit-learn and XGBoost.
           Computes SHAP values; saves model and artifacts for FastAPI.
NOTE     : ALL DATA IS SYNTHETIC. For demonstration purposes only.
"""

import sys
import os
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import seaborn as sns
import shap
import joblib

from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.metrics import (
    roc_auc_score, accuracy_score, f1_score,
    classification_report, roc_curve, ConfusionMatrixDisplay
)
from xgboost import XGBClassifier

warnings.filterwarnings("ignore")

# ============================================================
# CONFIG
# ============================================================

RANDOM_STATE = 42
MODEL_DIR    = Path("ml")
DATA_PATH    = Path("data/adam/adsl.csv")

FEATURE_COLS = [
    "AGE", "SEXN", "RACEN", "REGIONN",
    "DAS28BL", "BMIBL", "TRT01PN",
    "AENUM", "AESNUM", "AEMAXSEVN", "N_ALT_AE", "N_NEURO_GI_AE",
    "COMPLPCT", "N_MISSED_DOSES"
]
TARGET_COL = "DPTN"   # 1=dropout, 0=completed

NUMERIC_FEATURES     = ["AGE", "DAS28BL", "BMIBL", "COMPLPCT", "AENUM",
                         "AESNUM", "AEMAXSEVN", "N_ALT_AE", "N_NEURO_GI_AE", "N_MISSED_DOSES"]
CATEGORICAL_FEATURES = ["SEXN", "RACEN", "REGIONN", "TRT01PN"]

# ============================================================
# LOAD DATA
# ============================================================

print("=" * 60)
print("NovaStat NSVT-001 - Python ML Training")
print("NOTE: All data is AI-generated - for demonstration only")
print("=" * 60)

if not DATA_PATH.exists():
    sys.exit(
        f"ERROR: {DATA_PATH} not found.\n"
        "Run sas/generate_sdtm.sas and sas/create_adam.sas first."
    )

df = pd.read_csv(DATA_PATH)
print(f"\nLoaded ADSL: {len(df)} subjects | "
      f"Dropout rate: {df['DPTFL'].eq('Y').mean()*100:.1f}%")

# Prepare model data
model_df = (
    df[df["ITTFL"] == "Y"]
    .assign(DPTN=lambda x: (x["DPTFL"] == "Y").astype(int))
    [FEATURE_COLS + [TARGET_COL]]
    .dropna()
)
print(f"Model dataset: {len(model_df)} subjects | "
      f"Dropout rate: {model_df[TARGET_COL].mean()*100:.1f}%")

X = model_df[FEATURE_COLS]
y = model_df[TARGET_COL]

# ============================================================
# TRAIN / TEST SPLIT
# ============================================================

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=RANDOM_STATE, stratify=y
)
print(f"\nTrain: {len(X_train)} | Test: {len(X_test)} | "
      f"Train dropout rate: {y_train.mean()*100:.1f}%")

# ============================================================
# PREPROCESSING PIPELINE
# ============================================================

numeric_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler",  StandardScaler()),
])

categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
])

preprocessor = ColumnTransformer([
    ("num", numeric_pipeline,     NUMERIC_FEATURES),
    ("cat", categorical_pipeline, CATEGORICAL_FEATURES),
])

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

# ============================================================
# MODEL 1: RANDOM FOREST
# ============================================================

print("\nTraining Random Forest (GridSearch CV=5)...")
rf_pipe = Pipeline([
    ("prep", preprocessor),
    ("clf",  RandomForestClassifier(random_state=RANDOM_STATE, n_jobs=-1, class_weight="balanced"))
])

rf_grid = GridSearchCV(
    rf_pipe,
    param_grid={"clf__n_estimators": [200, 400], "clf__max_depth": [4, 6, None],
                "clf__min_samples_leaf": [5, 10]},
    cv=cv, scoring="roc_auc", n_jobs=-1, verbose=0
)
rf_grid.fit(X_train, y_train)

rf_best  = rf_grid.best_estimator_
rf_probs = rf_best.predict_proba(X_test)[:, 1]
rf_preds = rf_best.predict(X_test)

rf_metrics = {
    "model":    "Random Forest",
    "roc_auc":  round(roc_auc_score(y_test, rf_probs), 4),
    "accuracy": round(accuracy_score(y_test, rf_preds), 4),
    "f1":       round(f1_score(y_test, rf_preds), 4),
}
print(f"  RF  AUC={rf_metrics['roc_auc']:.3f} | ACC={rf_metrics['accuracy']:.3f} | "
      f"F1={rf_metrics['f1']:.3f}")

# ============================================================
# MODEL 2: XGBOOST
# ============================================================

print("\nTraining XGBoost (GridSearch CV=5)...")
xgb_pipe = Pipeline([
    ("prep", preprocessor),
    ("clf",  XGBClassifier(random_state=RANDOM_STATE, eval_metric="logloss",
                           use_label_encoder=False, n_jobs=-1))
])

xgb_grid = GridSearchCV(
    xgb_pipe,
    param_grid={"clf__n_estimators": [200, 400], "clf__max_depth": [3, 5, 7],
                "clf__learning_rate": [0.05, 0.1], "clf__subsample": [0.8, 1.0]},
    cv=cv, scoring="roc_auc", n_jobs=-1, verbose=0
)
xgb_grid.fit(X_train, y_train)

xgb_best  = xgb_grid.best_estimator_
xgb_probs = xgb_best.predict_proba(X_test)[:, 1]
xgb_preds = xgb_best.predict(X_test)

xgb_metrics = {
    "model":    "XGBoost",
    "roc_auc":  round(roc_auc_score(y_test, xgb_probs), 4),
    "accuracy": round(accuracy_score(y_test, xgb_preds), 4),
    "f1":       round(f1_score(y_test, xgb_preds), 4),
}
print(f"  XGB AUC={xgb_metrics['roc_auc']:.3f} | ACC={xgb_metrics['accuracy']:.3f} | "
      f"F1={xgb_metrics['f1']:.3f}")

# ============================================================
# SELECT BEST MODEL
# ============================================================

best_metrics = max([rf_metrics, xgb_metrics], key=lambda m: m["roc_auc"])
best_model   = rf_best if best_metrics["model"] == "Random Forest" else xgb_best
best_probs   = rf_probs if best_metrics["model"] == "Random Forest" else xgb_probs

print(f"\nBest model: {best_metrics['model']} (AUC = {best_metrics['roc_auc']:.3f})")

# ============================================================
# SHAP VALUES (on test set sample)
# ============================================================

print("\nComputing SHAP values...")

# Get preprocessed test data for SHAP
X_test_prep = best_model.named_steps["prep"].transform(X_test)
feature_names_out = (
    NUMERIC_FEATURES
    + CATEGORICAL_FEATURES  # after passthrough
)

try:
    if best_metrics["model"] == "XGBoost":
        explainer    = shap.TreeExplainer(best_model.named_steps["clf"])
        shap_values  = explainer.shap_values(X_test_prep)
        expected_val = explainer.expected_value
    else:
        explainer    = shap.TreeExplainer(best_model.named_steps["clf"])
        shap_values  = explainer.shap_values(X_test_prep)[1]  # class 1 (dropout)
        expected_val = explainer.expected_value[1]

    shap_computed = True
    print("  SHAP values computed successfully.")
except Exception as e:
    shap_computed = False
    shap_values   = None
    expected_val  = None
    print(f"  SHAP skipped: {e}")

# ============================================================
# SAVE ARTIFACTS
# ============================================================

MODEL_DIR.mkdir(exist_ok=True)

joblib.dump(best_model,   MODEL_DIR / "dropout_model_py.pkl")
joblib.dump(preprocessor, MODEL_DIR / "preprocessor.pkl")

# Save metadata
metadata = {
    "study":        "NSVT-001",
    "sponsor":      "NovaStat Therapeutics",
    "drug":         "NST-4892",
    "target":       "Patient dropout (DPTN=1)",
    "best_model":   best_metrics["model"],
    "roc_auc":      best_metrics["roc_auc"],
    "accuracy":     best_metrics["accuracy"],
    "f1":           best_metrics["f1"],
    "features":     FEATURE_COLS,
    "n_train":      len(X_train),
    "n_test":       len(X_test),
    "rf_metrics":   rf_metrics,
    "xgb_metrics":  xgb_metrics,
    "disclaimer":   "SYNTHETIC DATA - AI-generated for demonstration only",
}
joblib.dump(metadata, MODEL_DIR / "model_metadata_py.pkl")

if shap_computed:
    joblib.dump({
        "shap_values":   shap_values,
        "expected_value": expected_val,
        "feature_names": feature_names_out,
        "X_test_prep":   X_test_prep,
    }, MODEL_DIR / "shap_artifacts.pkl")
    print(f"SHAP artifacts saved to {MODEL_DIR}/shap_artifacts.pkl")

print(f"\nModel saved to {MODEL_DIR}/dropout_model_py.pkl")

# ============================================================
# PRINT CLASSIFICATION REPORT
# ============================================================

print(f"\n--- {best_metrics['model']} Classification Report (Test Set) ---")
print(classification_report(
    y_test, best_model.predict(X_test),
    target_names=["Completed", "Dropout"]
))

print("Training complete.")
print("Start the API with: cd novastat_pharma-csp && uv run uvicorn python.api:app --reload")
